### PR TITLE
feat(code): Red Alert visual warning

### PR DESCRIPTION
**Pull request by Zitchas**
**Feature:** This PR implements the feature request detailed and discussed in issue #7766 

## Feature Details
This implements a flashing red warning whenever the Red Alert siren goes off in the upper left corner in the wasted space above and to the left of the radar. 

This is intended to provide greater accessibility for a variety of people:
a) people who are hard of hearing and have difficulty hearing the red alert siren.
b) people playing on computers that have malfunctioning or absent speakers.
c) people playing in environments where sound is inadvisable.

## Credit

Thank you to VitalChip for providing the code for this. I slightly modified the timing, and the image is entirely mine, but all the code proper is theirs.

## UI Screenshots

Red Alert in action

## Testing Done
Played the game. Went somewhere where there are hostiles, observed the alert flashing.

## Performance Impact
Unknown. Presumably fairly minor, as it is simply flashing a single tiny icon. No visible impact on the display or speed of gameplay.

## Artwork
The `red alert.png` is entirely my own creation, and not derivative of anything else. The base is created in LibreOffice Draw, then exported to GIMP for resizing and adding the RADAR hit dots.

A zip file on google drive with the LibreOffice Draw file, the GIMP file, and both regular and x2 copies of the .png can be [downloaded here](https://drive.google.com/file/d/14fIMLIDN_kSQOl76da4W6d8gkLOIxU_y/view?usp=share_link) and will be PR'd to the ES asset repository momentarily.

### CODE REVIEW

**Comment by warp-core** [comment]
Why not have `Engine::Step()` set a condition in the interface info when the sprite is to be displayed and have the interface file just define the sprite with a `visible if <the condition you set in Engine::Step()>`?

**Comment by Zitchas** [comment]
@vitalchip Could you answer warp-core's question please? I don't know the answer.

**Comment by vitalchip** [comment]
Nothing more than personal preference.

Typically I set a condition if I want a backgroung graphic or a button to appear under certain conditions but for an indicator I set the sprite/info.

If I decided later that I also want a yellow alert symbol I'd just conditionally push a different sprite to "alarm sprite" and I wouldn't haave to go back to interface.txt to create a second "visible if"  to pickup a second SetCondition So it's easier to play around with.

Maybe I just like the easy life ;-)

**Comment by quyykk** [comment]
The red is a bit jarring IMO. Maybe make it a lighter red?

**Comment by Zitchas** [comment]
Agreed, however it's a red alert warning indicator. It's *supposed* to be jarring, aggressive, and demanding.

**Comment by RisingLeaf** [comment]
I don't want to make this sound bad, but I think it would be much easier for everyone if code PR's would be opened by actual coders.

I don't mind it for small things like universal ramscoop, but adding a flashing label is a bit more complex and in my opinion would be better if the coder had direct access to the code, if you get what I mean.

**Comment by RisingLeaf** [comment]
That said i still appreciate your efforts Zitchas

**Comment by Zitchas** [comment]
a) It helps me learn the code, and
b) VitalChip seems.. reluctant... to do PRs right now but is helpfully providing working code.
c) No-one else is working on it. 
d) Is there actually a problem that you're not mentioning, or just an "in general" thing?

**Comment by Saugia** [comment]
I'd appreciate if this was done as a toggleable, personally I would never have this because I don't like my UI flooded, but I see the reasons for the purposes described.

File: data/interfaces.txt [hunk 1]
```diff
@@ -527,6 +527,10 @@ interface "hud"
 
 	# Targets.
 	anchor top left
+	visible if "red alert"
+	sprite "ui/red alert"
+		from 10 10
+		align top left
```
Comment (warp-core) [inline data/interfaces.txt:533]:
```suggestion
		align top left
	visible
```

**Comment by RisingLeaf** [comment]
> a) It helps me learn the code, and
> 
> b) VitalChip seems.. reluctant... to do PRs right now but is helpfully providing working code.
> 
> c) No-one else is working on it. 
> 
> d) Is there actually a problem that you're not mentioning, or just an "in general" thing?
> 
> 
> 
> 
> 
> This isn't a "fired and forgotten" code PR, I am doing my best to make changes as requested and sometimes, yeah, that is going and getting new code from the main coder; but nothing's impairing your ability to suggest code.
> 
> 
> 
> edit2: Is there anything that would be going differently or better if this was Vitalchip's PR rather than mine?

Nah, it just slows down the process, it's fine

**Review by warp-core — CHANGES_REQUESTED** [review]
The `visible if` is applied to everything after it until it is override with either another `visible if` or a lone `visible` (which removed any condition requirement) or this interface definition ends.

**Comment by Hurleveur** [comment]
We could also have the siren option changed to a "warning" setting that goes from none to visual to sound to both?

**Comment by Zireael07** [comment]
+1 to what @Hurleveur suggests (I'd prefer not to annoy my flatmates if I forget to switch speakers off - they're at max loudness when on, for reasons)

File: source/Engine.cpp [hunk 2]
```diff
@@ -679,6 +679,8 @@ void Engine::Step(bool isActive)
 	info.SetString("date", player.GetDate().ToString());
 	if(flagship)
 	{
+		if(alarmTime && step / 20 % 2)
```
Comment (RisingLeaf) [inline source/Engine.cpp:682]:
```suggestion
		// Have an alarm label flash up when enemy ships are in the system
		if(alarmTime && step / 20 % 2)
```

**Review by RisingLeaf — APPROVED** [review]
Looks structural good

**Review by RisingLeaf — APPROVED** [review]
Looks good to me

**Review by warp-core — COMMENTED** [review]
Perhaps you didn't see the addition to my earlier comment:

> Edit: and I agree with Saugia that this should be optional (the audible siren can be disabled in the settings.)

> You'll need to add an entry for the option to the SETTINGS array in PreferencesPanel::DrawSettings() (I think this is the method name) and add && Preferences::Get("your setting name") to the if statement you've added in Engine:;Step. You'll also need to add it to the default values in Preferences.cpp so it is on to begin with (like the audible siren).

Hurleveur has also proposed integrating an option for this into the existing "warning siren" preference. That would probably then be renamed to something else, and would have four options (instead of the current `on` and `off`): `off`, `audible`, `visual`, `both` (these are my suggestions for the names, anyone is free to suggest alternatives).
Doing this is a bit more complicated. Off the top of my head, you'll need an enum class and explicit functions for getting and setting this preference in the Preferences class, and an exception when drawing the settings in `PreferencesPanel::DrawSettings` so that it writes the appropriate string instead of `on` or `off`. See the boarding priorities PR for an example, there are also some existing settings that use a similar arrangement, though I don't think any are quite as complex as this.
This would avoid the issue of the two preferences for the different red alert indicators being in totally separate parts of the settings panel.

If you are happy with this design (single preference controlling both), I might be able to open a PR to your branch later today.

**Comment by Zitchas** [comment]
Sorry, I took the earlier post as an idea, not a requirement. It's kind of outside the scope of just adding a visual to the red alert siren for accessibility purposes, but sure, if you want to PR some code to my branch to add in the preference option for "off, audible, visual, both", that'd be great.

**Comment by warp-core** [comment]
> Sorry, I took the earlier post as an idea, not a requirement. It's kind of outside the scope of just adding a visual to the red alert siren for accessibility purposes, but sure, if you want to PR some code to my branch to add in the preference option for "off, audible, visual, both", that'd be great.

It's a requirement for my approval, but others might not feel the same way I do.

**Comment by Zitchas** [comment]
Ah. Ok. Well, yes, having one setting controlling both (via the off, both, audible, visual options) sounds like the most effective way of doing it.

**Comment by Zitchas** [comment]
Thanks to warp-core for PR'ing a preference setting. I tested it partially, and the visual indicator does come up under "both" and "visual" settings. I don't have sound right now to check the audio side.

**Review by Amazinite — COMMENTED** [review]
I think the icon should use the same red color as when you're targeted by an enemy on the minimap. That is, something much less harsh.

**Comment by Zitchas** [comment]
As I commented before, the color choice is intentionally harsh because it's supposed to grab the eye and trigger that danger/excitement/something changed! reflex. I don't think that it should be blending into other colors that may be present.

**Comment by Hurleveur** [comment]
I think that making it flash would bring attention whilst being able to lower that red?

**Comment by Zireael07** [comment]
@Hurleveur: Flashing can trigger migraines/epilepsy in some people, so IMHO **not** a good idea.

**Comment by Hurleveur** [comment]
Well I wouldn't mean instant bright flashing, more like fading in an out nicely?

**Comment by Zitchas** [comment]
Personally, I find flashing is a means to ensure something *keeps* a person's attention, but isn't so good at *getting* attention. A bright, harsh color is better at getting that attention the instant it appears, without waiting for an animation to generate action & draw the attention. 

We're talking about something whose purpose is to inform people that there is an immediate life-threatening critical situation. This isn't supposed to look pretty, it definitely isn't supposed to blend in or match color schemes. It's supposed to be aggressive, demanding attention, and pulling it in the fastest, most instantaneous method possible. 

Now, there's an argument to be made that after the initial appearance that it fading down a bit might be a reasonable thing to do - but typically it's gone after a few seconds anyway, so that seems like a extra effort for very little gain.

**Comment by Zitchas** [comment]
The image is now slightly darker. Hopefully this is an acceptable compromise brightness.

**Review by Hurleveur — APPROVED** [review]
Codewise this looks good, and the image seems fine to me

**Comment by warp-core** [comment]
[Zitchas#126](https://github.com/Zitchas/endless-sky/pull/126)

**Comment by warp-core** [comment]
What are we calling the preference?
It's currently `Alert warning`.
Does anyone have a preference?
~~I've used `Alert indicator` internally which should probably be changed to match the name in the settings panel.~~

**Comment by Hurleveur** [comment]
Alert indicator sounds good

**Comment by Zitchas** [comment]
I think having it displayed as "Alert indicator" is a little clearer in identifying what it is (the indicator that the alert is happening); and has the bonus of matching the internal labels.

**Comment by Amazinite** [comment]
It's not a necessity for the icon to be pure red in order to catch the eye. Its very presence when normally it'd be missing is very eye catching already. I'm just asking to use the same color we currently use to denote hostiles on the minimap (`color "radar hostile" 1. .6 .4 0.`) for the sake of consistency across UI elements.
red alert icon

**Comment by Zitchas** [comment]
It might just be my eyes, but that other color is more orange than red to me. Given the choice, I'd actually prefer to change it to being a solid red, so that the "hostile but not focused on me" color can be more orange and thus not blend in with the color of stars and the ship I have selected.

**Comment by Hurleveur** [comment]
I see it as orange too but I think its fine as orange anyway? Isnt there some color issue with red for some ppl?

**Comment by Zitchas** [comment]
Some info on colorblind color choices:
https://michigansignshops.com/blog/signs-for-colorblind-viewers/

The very short summary is:
- monochrome is better
- The higher the contrast, the better.
- There doesn't appear to be a problem with red by itself, but there *is* a problem with red in conjunction with green

<details>

<summary>List of color combinations to avoid</summary>

- Green and red
- Brown and green
- Blue and purple
- Green and blue
- Light green and yellow
- Blue and grey
- Green and grey
- Black and green

</details>

**Comment by Amazinite** [comment]
I think the important thing I'm looking for is consistency, so if we also change the default hostile radar color then that's fine, although I think pure red still isn't the best. How about `1. .3 .2`?
image

**Comment by Hurleveur** [comment]
I still prefer orange honestly its less agressive and still gets the message accross? I think I saw somewhere red was excluded from UI on purpose but idk what issue it was.
That said a general alert could be reder than just an enemy targeting you?

**Comment by Zitchas** [comment]
I'll look into making the colors for the two consistent, but won't be able to immediately.

That being said, is having the color of the warning match the color of the hostiles really a "merge blocking" detail for this functionality? I'm willing to do it just to make one person happy, but it really doesn't seem like something that is even a given as being something that is forcibly necessary to begin with. 

On a different topic, colors in UX for accessibility purposes.
Another interesting link to an article on colors in UI here:
https://uxplanet.org/using-red-and-green-in-ui-design-66b39e13de91?gi=da6de78f037e

Colors for Accessibility

**Comment by Zireael07** [comment]
> That being said, is having the color of the warning match the color of the hostiles really a "merge blocking" detail for this functionality?

IMHO not, and as a person who (indirectly) was the trigger to making this PR, I would like this to be merged sooner rather than later as it'll likely alleviate an issue I wasn't even aware I had (me not noticing I'm in danger was likely a factor in all three of my attempts at playing getting abandoned really fast)

**Comment by Amazinite** [comment]
The next release isn't for another month and a half at least. This isn't some critical feature that needs added immediately. We can deal with the color now.

**Comment by Saugia** [comment]
Coming from an artistic perspective, I have to agree with Derpy on the colors matching. What's the purpose of them having separate colors? If they are separate, that leaves room for question as to why the red warning icon is not the same as the ships... are the ships not as threatening as the red suggests on the icon warning, etc..The ships being hostile (the orange) should be a little more red, and the icon should use that same red, if we are using this route.

**Comment by Saugia** [comment]
Also, not sure how I feel about the icon in general, especially the location. What's the purpose of putting the warning icon *right next* to the radar? If you're going to be looking at that part of the screen anyway at the radar, it's redundant, since you will see the red dots on the radar at the same time. I either suggest moving the icon somewhere else on the UI not right next to the radar, or foregoing an icon entirely and using a different means of alert.

**Comment by Zitchas** [comment]
There isn't really a purpose in being separate colors, just I don't see a strong reason to insist that they are the same. I'm fine with doing so, though.

Well, that's one reason why it's a big, brightly colored, flashing symbol where normally there is nothing. Zero chance of being glossed over, extremely high contrast with surroundings; and serves to draw the eye to the radar corner which is where the player would likely get the most useful information about what is going on. So Eye goes to the  bright flashy thing, recognize that it is a warning with a radar icon, and one's eye is already right there in the general vicinity of the radar to identify where the hostiles are. 

It's not really a useful warning to someone already watching the minimap, since they'll already know. But it *is* a means to get people to look that way.

**Comment by 1010todd** [comment]
Having the alert icon next to the radar seems to be the most fitting position for me, Bottom left is your fleet ships, bottom right is secondaries, top right is info about you, top left is the minimap, the stuffs around you. I agree with Zitchas it'll draw attention to the minimap for those who aren't already looking and now they'll have a glance at it giving some idea where hostile is. 

For the color, I kinda agree with Derpy it should be consistent with hostile ship color. I don't have strong feeling about which color it should be but people seems to like contrast I guess.

**Comment by Zireael07** [comment]
@1010todd good point, people look at the minimap in combat anyway

**Comment by Zitchas** [comment]
Just to expound on the location selection a bit more so people understand that the top left wasn't just a "well, this is easiest" choice, but rather a deliberate "this is the best place to put it" decision.


Upper right is info about your flagship. Shield & Hull amounts, fuel, energy, and heat levels. If I saw a warning in this area, I'd expect it to be something related to my ship's performance. A "low fuel" warning, or perhaps a "critical hull failure imminent" or a "power failure" warning. Warnings about being overheated, or about being slowed or disrupted or subject to various other effects that change my ship's behavior should probably have warnings or displays up here of some kind. Being under-crewed might deserve a warning here too.

Bottom right is secondary weapons and ammo. So I'd expect this area to have warnings about running out of ammo, or weapon system failures. As far as that goes, might be a good place to put a warning about ionization, since that affects weapon systems.

In both these areas, my first thought upon seeing a warning image would be "something's wrong with my ship or systems. Everything over here is internals, after all. 

The top of the screen is currently un-used. So is the left edge, and to some extent the bottom edge. And, obviously, there's the whole central area. 

Now, the central area is where the action is. It's where we see our ship, where all the gunfire and missiles and dodging and everything else takes place. Having an alert icon (or a "Red Alert" warning, or whatever else) pop up here would certainly be the most attention grabbing option we could pick, but it would also obscure our ship and all the action that's actually happening. Given how short our response times are already, not to mention how short all our ranges are, obscuring the main action area for even fractions of a second seems both counterproductive and deliberately harmful to the player.

The center of the top, left, and bottom edges are basically empty voids. There's nothing there (although hails sometimes encroach into the bottom edge.) So there's no pre-determined use for the areas to conflict with. On the flip side, though, there's nothing to provide context to an alert, either. This means our icon has to be better, or perhaps forgo an icon entirely for literally spelling out something like `[ WARNING: HOSTILES PRESENT ]`. That's very explicit, which is good. It is also in English, which means that any attempt to translate the game in the future will have to concern itself not only with the text of the game, but also creating new artwork, too. It also requires reading, which takes longer, and time is of the essence here. Assuming we just use an icon, from a usability standpoint, there's zero chance of it blending into antyhing, which is a good thing. But it will draw the eye to here, tell the player there's hostiles, and then we have to go over to the radar to see where they are. So this generates travel time. Eye travel time might be fast, but it's still extra time being added to the response time sequence. In the case of the bottom or right edges, this is actually drawing the eye to almost the furthest point on the screen from where they will be getting additional information about the warning.

Operational areas:

For most people, the operational field of view is a circle around their ship. Personally, this circle is a little longer than my longest ranged primary weapon (or a little longer than the longest range of the primary weapon I expect to go up against.) This area covers everything that is the most important in terms of the actions: our flagship, the ships we're fighting, whatever we're mining, etc. Ideally, this area should occupy the largest portion of our screen possible. For most people, this will be the vertical height of their screen. For some people, it will be the width (people with screens in portrait mode). In general, I think we should keep UI elements out of the center-edge area because that impinges on and limits how zoomed in they can get and still see the entirety of their "operational zone" around their ship. 

In contrast, the corners of the screen are the areas that are furthest from our flagship, and thus the areas where we can put stuff with the least impact on what the player can see going on around them. ES already follows this, but putting virtually all of our in-flight UI into the corners.

In this particular case, we're putting it above and to the left of the minimap. This area - while empty - is already basically a "dead" area of the screen because it is blocked off from the rest of the screen by the minimap and targeting info. Thus using it for the alert symbol is making use of an area that is already basically useless for seeing the on-screen action, and doesn't encroach on the area the player can use to see what's going on around their ship.

## Summary:

So, while they are all options, my personal judgement is that the top left makes for the best location:
- That corner is already the "externals" corner. All info presented here is related to the external environment around the ship, so a warning here is already in the context of "alert about something outside our ship, in the area around us." So even if the player has no clue what the icon means, just by its placement they can have a pretty good idea that it has something to do with the things around them.
- On the off chance that the player doesn't use the minimap, the alert will draw their attention to that corner and help them to realize the valuable information it presents.
- Once the player knows there are hostiles present, the next logical source of information is either a) looking at the mini map,or b) hitting the "nearest hostile" button and looking at the targeting info. In both cases, these areas are immediately adjacent, also in the top left corner. This minimizes eye transit time, as the player is already looking at that corner.
- Refrains from impacting the in-flight operational visibility area.

**Comment by Zitchas** [comment]
Other options I've seen in other games:
- Screen highlight: When a hostile is present, the entire edge of the screen glows/pulses red. In at least one game it's just permanently a red border so long as hostiles are present. Does have the problem of effectively shrinking the screen a bit.
- Flagship red highlight: When the red alert is triggered, the flagship pulses red briefly. Very effectively gets the players attention, but also makes it hard to see any effects or anything about the flagship and what's going on with it.
- Flashing red border on the mini-map. Perhaps my second favorite option, after just having a warning icon outright. Draws the attention to the mini-map, in effect making it the warning. This has visibility problems for people who have color problems, since for them it may just look like the minimap is pulsing the same color as it is normally, which won't stand out very much. Could be effective if the border actually had some kind of change to it so it wasn't just a color & pulse, but an actual shape change, (say, going from plain circle to bold circle with targeting type brackets around it or something, although for this to be effective it'd need to be permanent so long as the hostiles are present, probably)

**Comment by Zireael07** [comment]
IMHO the icon is a good idea for now - if there are complaints or wishes for one of the border/minimap ideas, they can be done in a follow up PR

**Comment by Zitchas** [comment]
Alright, the color that Amazinite suggested is, in my opinion, significantly better than the current hostile color, so I have changed it as well as the icon to match. So all the hostile stuff should be the same color now.

File: data/interfaces.txt [hunk 3]
```diff
@@ -99,7 +99,7 @@ color "missile dangerous" 1. .1 .1 1.
 color "radar player" .2 1. 0. 0.
 color "radar friendly" .4 .6 1. 0.
 color "radar unfriendly" .8 .8 .4 0.
-color "radar hostile" 1. .6 .4 0.
+color "radar hostile" 1. .3 .2 0.
```
Comment (RisingLeaf) [inline data/interfaces.txt:102]: seems like an unrelated change?
-> separate PR

File: source/Preferences.cpp [hunk 4]
```diff
@@ -51,6 +51,9 @@ namespace {
 
 	const vector<string> BOARDING_SETTINGS = {"proximity", "value", "mixed"};
 	int boardingIndex = 0;
+
+	const vector<string> ALERT_INDICATOR_SETTING = {"off", "audio", "visual", "both"};
+	int alertIndicatorIndex = 3;
```
Comment (RisingLeaf) [inline source/Preferences.cpp:56]: should it really start with both?

File: source/Preferences.cpp [hunk 5]
```diff
@@ -96,9 +98,21 @@ void Preferences::Load()
 			vsyncIndex = max<int>(0, min<int>(node.Value(1), VSYNC_SETTINGS.size() - 1));
 		else if(node.Token(0) == "fullscreen")
 			screenModeIndex = max<int>(0, min<int>(node.Value(1), SCREEN_MODE_SETTINGS.size() - 1));
+		else if(node.Token(0) == "alert indicator")
+			alertIndicatorIndex = max<int>(0, min<int>(node.Value(1), ALERT_INDICATOR_SETTING.size()));
```
Comment (RisingLeaf) [inline source/Preferences.cpp:102]:
```suggestion
			alertIndicatorIndex = max<int>(0, min<int>(node.Value(1), ALERT_INDICATOR_SETTING.size() - 1));
```

Comment (RisingLeaf) [reply source/Preferences.cpp:102]: some memory errors otherwise

File: source/PreferencesPanel.cpp [hunk 6]
```diff
@@ -506,7 +509,7 @@ void PreferencesPanel::DrawSettings()
 		"Show escort systems on map",
 		"Show stored outfits on map",
 		"System map sends move orders",
-		"Warning siren"
+		ALERT_INDICATOR
```
Comment (RisingLeaf) [inline source/PreferencesPanel.cpp:512]: visual stuff is further up

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
Some small stuff

File: data/interfaces.txt (hunk 3 above)
Comment (warp-core) [reply data/interfaces.txt:102]: This is done to ensure that the new icon and the colour for hostile ships on the RADAR match.
Doing it in a separate PR seems like a waste of time.

File: source/Preferences.cpp (hunk 4 above)
Comment (warp-core) [reply source/Preferences.cpp:56]: Yes. The current default is to have the audible siren on, I don't see why we wouldn't also have this indicator on.

Comment (RisingLeaf) [reply source/Preferences.cpp:56]: ok

File: data/interfaces.txt (hunk 3 above)
Comment (RisingLeaf) [reply data/interfaces.txt:102]: ah, didnt see that, fine then

File: source/PreferencesPanel.cpp (hunk 6 above)
Comment (warp-core) [reply source/PreferencesPanel.cpp:512]: This setting is not exclusively display related.

File: source/Preferences.cpp (hunk 5 above)
Comment (warp-core) [reply source/Preferences.cpp:102]: @Zitchas can you apply this suggestion?

File: data/interfaces.txt (hunk 3 above)
Comment (Amazinite) [inline data/interfaces.txt:102]:
```suggestion
color "radar hostile" .8 .3 .2 0.
```

Comment (Amazinite) [inline data/interfaces.txt:102]: Messing around with this, I think this new color is a tad bit too bright. Turns out this color is used for more than just the radar; the targeted ship outline, targeted ship blips, and targeted planet blips also use this color. (And the hostile color is used for restricted planets that you have selected, which was a bit surprising.) While a bright color would work on just the radar, having this color used elsewhere was a bit jarring to me, so bringing the red component down a bit to lower the overall brightness. Those other uses of this color could potentially be split off, but that shouldn't be the responsibility of this PR.

I'm fine with the icon being slightly brighter and keeping its current color.
```suggestion
color "radar hostile" .85 .3 .2 0.
```

**Comment by Zireael07** [comment]
Thank you, thank you!
